ic-stable-structures/src/memory_mapped_files/memory_mapped_file.rs

```diff
@@ -0,0 +1,358 @@
+use std::{
+    fs::{copy, remove_file, File, OpenOptions},
+    path::Path,
+};
+
+use memmap2::{MmapMut, MmapOptions};
+
+use super::error::{MemMapError, MemMapResult};
+
+/// By default we use chunk size equal to the default page size.
+/// Since our structures are usually pretty small it doesn't seem
+/// that we will benefit from using huge page size (2 MB or 1 GB)
+const PAGE_SIZE: u64 = 4096;
+
+/// When creating mapping we reserve at once 1 TB of address space.
+/// This doesn't allocate any resources (except of address space which is not a problem for x64)
+/// but allows skip remapping/flushing when the file size grows.
+const MEM_MAP_RESERVED_LENGTH: u64 = 1 << 40;
+
+/// Memory mapped file implementation.
+/// If `is_permanent` flag is true then after the
+/// structure is dropped all the changes are saved to file.
+pub(super) struct MemoryMappedFile {
+    file: File,
+    path: String,
+    length: u64,
+    is_permanent: bool,
+    mapping: MmapMut,
+}
+
+impl MemoryMappedFile {
+    /// Preconditions: file under the `path` should not be modified from any other place
+    /// in this or different process.
+    pub fn new(path: String, is_permanent: bool) -> MemMapResult<Self> {
+        if !is_permanent {
+            _ = remove_file(&path);
+        }
+
+        let file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .write(true)
+            .read(true)
+            .open(&path)?;
+        let length = file.metadata()?.len();
+
+        let mut mmap_opts = MmapOptions::new();
+        // Safety: function preconditions should guarantee the safetry of the operation:
+        // mapping to a file is safe if the file isn't modified concurrently by this and other processes.
```

Comment:
what about thread concurrency?

Reply:
As it is said in the function description - it is unsafe to have a different mapping for the same file no matter in this or other thread.
But it is safe to use the same mapping from a different thread if no concurrent writes or read-writes happen. But because we have `&self` and `&mut sefl` parameters for read and write operations correspondingly this will be guaranteed by the Rust borrow checker